From 86900e65a4e5f60c9e156ea568eb07c61740e951 Mon Sep 17 00:00:00 2001
From: Ben Peled <bpeled@nvidia.com>
Date: Tue, 21 Nov 2023 06:15:45 -0800
Subject: [PATCH] init hold lock before registering devices

some devices register includes i2c transfer
i2c transfer uses hold lock
this creates situation were lock is taken before init

Signed-off-by: Ben Peled <bpeled@nvidia.com>
Upstream-Status: Pending [Not submitted to upstream yet]
---
 drivers/i2c/i2c-core-base.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/i2c/i2c-core-base.c b/drivers/i2c/i2c-core-base.c
index 8bdb64ff8048..1a249a3b77da 100644
--- a/drivers/i2c/i2c-core-base.c
+++ b/drivers/i2c/i2c-core-base.c
@@ -1598,6 +1598,7 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 
 	dev_dbg(&adap->dev, "adapter [%s] registered\n", adap->name);
 
+	mutex_init(&adap->hold_lock);
 	/* create pre-declared device nodes */
 	of_i2c_register_devices(adap);
 	i2c_acpi_install_space_handler(adap);
@@ -1611,7 +1612,6 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 	bus_for_each_drv(&i2c_bus_type, NULL, adap, __process_new_adapter);
 	mutex_unlock(&core_lock);
 
-	mutex_init(&adap->hold_lock);
 	INIT_DELAYED_WORK(&adap->unhold_work, i2c_adapter_unhold_work);
 
 	return 0;
-- 
2.43.0

